app/controllers/concerns/eventable.rb

```diff
@@ -0,0 +1,38 @@
+module Eventable
+  extend ActiveSupport::Concern
+
+  def supplier_id
+    # NB: not all events will have a supplier_id so this could well be nil
+    @supplier_id ||= current_user.owner&.id
+  end
+
+  def process_event(eventables, event_name, event_params)
+    [eventables].flatten.each do |eventable|
+      event = create_event(eventable, event_name, event_params)
+      create_generic_event(event)
+      run_event_logs(eventable)
+    end
+  end
+
+  def create_event(eventable, event_name, event_params)
+    eventable.events.create!(
+      event_name: event_name,
+      client_timestamp: Time.zone.parse(event_params.dig(:attributes, :timestamp)),
+      details: {
+        event_params: event_params,
+        supplier_id: supplier_id,
+      },
+    )
+  end
+
+  def create_generic_event(event)
+    generic_event = GenericEvent.from_event(event)
```

Comment:
creating the generic event from the v1-event is a bit convoluted and means the event needs to be updated after its been saved. How about instead creating the generic event first, and then creating the v1-event referencing the generic event second?

Reply:
This is an old pattern we're following and that I plan to refactor later when I come to remove dual writing.

Reply:
Explained this offline and we agreed that dual-writing is technically not needed anymore, but that because updating the `#from_event` builders is going to create quite a lot of PR noise this should be done separately. I'm going to prioritise that next.
 
1. I wanted a consistent approach to running event logs between the old and the new apis in this project. This led to